ngcc: implement ivy switch rendering

petebacondarwin opened the pull request:
This is related to #25238

petebacondarwin committed:
test(ivy): refactor Esm2015Renderer tests to make them less fragile

petebacondarwin committed:
feat(ivy): implement `Renderer.getSwitchableDeclarations`

This supports the "ngcc ivy switch" specified in #25238.

petebacondarwin committed:
feat(ivy): implement `NgccReflectionHost.getSwitchableDeclarations()`

This method will be used to find all the places where the "ivy switch"
will occur. See #25238

petebacondarwin commented:
One concern I have is whether there will be additional places that need updating outside of the `AnalyzedFile`s that get passed to the `Renderer`...  Any thoughts @alxhub / @mhevery  ?

mhevery commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -198,6 +199,16 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * @returns An array of variable declarations that match.
+   */
+  getSwitchableDeclarations(module: ts.Node): SwitchableVariableDeclaration[] {
+    return findAll(module, isSwitchableVariableDeclaration);
```
As of right now this switch is in `@angular/core` only. Should we short circuit this for perf reasons? Is there an argument to be made to enable such switch for others?

petebacondarwin replied:
I need to test this but I suspect that the bottleneck for performance is creating the TS compilation, which is necessary for the rest of the work. So I believe that parsing for these markers will make little performance difference.

petebacondarwin replied:
I suggest that we have a period of performance tuning once the functionality is in place. I think that there is potentially a lot of duplication of work, especially in the ngtsc resolver and the reflection hosts right now that would benefit from some caching strategies.

mhevery reviewed (approved):
There seems to be other unrelated changes in this PR, but the `__PRE_NGCC__` rewrite to `__POST_NGCC__` looks good to me.

petebacondarwin commented:
The other changes were to simplify the rendering tests, without which the changes to the tests would have got yet more complicated.

gkalpak commented on packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts:
```diff
@@ -27,74 +27,119 @@ function analyze(parser: Esm5FileParser, analyzer: Analyzer, file: ts.SourceFile
-        contents: `
+const PROGRAM = {
+  name: 'some/file.js',
+  contents: `
```
Using a helper function to strip indentation makes the code look (and get folded) much better. I usually have:

```ts
const stripIndentation = (input: string): string => {
  const lines = input.replace(/^ *\n/, '').replace(/\n *$/, '').split('\n');
  const minIndentation = Math.min(...lines.
    filter(l => !/^ *$/.test(l)).
    map(l => /^ */.exec(l)![0].length));
  const re = new RegExp(`^ {0,${minIndentation}}`);

  return lines.map(l => l.replace(re, '')).join('\n');
};
```

petebacondarwin replied:
Fair enough. This has been a worry for me for a while. I'll give this a go.

petebacondarwin replied:
why the two replaces? Blank lines get stripped in the filter right?

gkalpak replied:
SGTM.

> why the two replaces? Blank lines get stripped in the filter right?

The two `replace` remove leading/trailing empty/whitespace-only lines from the output. The `filter` is only used to ignore whitespace-only lines for computing the min indentation.

petebacondarwin replied:
RIght!

gkalpak commented on packages/compiler-cli/src/ngcc/test/rendering/esm5_renderer_spec.ts:
```diff
@@ -27,74 +27,119 @@ function analyze(parser: Esm5FileParser, analyzer: Analyzer, file: ts.SourceFile
   });
 
+  describe('rewriteSwitchableDeclarations', () => {
+    it('switch marked declaration initializers', () => {
```
+should (?)

gkalpak reviewed (approved):
Some minor comments. LGTM otherwise 👍

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/ngcc_host.ts:
```diff
@@ -8,9 +8,33 @@
+export const POST_NGCC_MARKER = '__POST_NGCC__';
+
+export type SwitchableVariableDeclaration = ts.VariableDeclaration & {initializer: ts.Identifier};
+export function isSwitchableVariableDeclaration(node: ts.Node):
```
We have lots of type constraint functions in the ngcc and ngtsc code. They are all standalone.

petebacondarwin commented on packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -353,6 +353,24 @@ const EXPORTS_FILES = [
+const MARKER_FILE = {
+  name: '/marker.js',
+  contents: `
+    let compileNgModuleFactory = compileNgModuleFactory__PRE_NGCC__;
```
Ooh. 👍

petebacondarwin committed:
test(ivy): refactor Esm5Renderer tests to make them less fragile

petebacondarwin committed:
test(ivy): test `Esm5Renderer.getSwitchableDeclarations`

Also incorporates a refactoring of the tests to make them less fragile.

petebacondarwin commented:
Rebased on master. Hoping @alxhub can get this through G3 etc.

mary-poppins commented:
You can preview ae1b36f at https://pr25534-ae1b36f.ngbuilds.io/.

alxhub commented on packages/compiler-cli/src/ngcc/test/rendering/renderer_spec.ts:
```diff
@@ -68,7 +71,7 @@ describe('Renderer', () => {
   });
   const RENDERED_CONTENTS =
-      `\n// REMOVE DECORATORS\n\n// ADD IMPORTS\n\n// ADD CONSTANTS\n\n// ADD DEFINITIONS\n` +
+      `\n// REWRITTEN DECLARATIONS\n\n// REMOVE DECORATORS\n\n// ADD IMPORTS\n\n// ADD CONSTANTS\n\n// ADD DEFINITIONS\n` +
```
I'm becoming less of a fan of this approach of testing. It's really hard to look at this test and figure out whether it's correct.

Can we start thinking of a way to make this more straightforward? It doesn't have to happen in this PR.

alxhub commented on packages/compiler-cli/src/ngcc/src/host/fesm2015_host.ts:
```diff
@@ -198,6 +198,16 @@ export class Fesm2015ReflectionHost extends TypeScriptReflectionHost implements
+   * @returns An array of variable declarations that match.
+   */
+  getSwitchableDeclarations(module: ts.Node): SwitchableVariableDeclaration[] {
+    return findAll(module, isSwitchableVariableDeclaration);
```
We should implement a performance optimization where we check module.getText() for `PRE_NGCC_MARER` via regex or `.indexOf`. Only if the test is positive should we walk the AST to actually find them.

petebacondarwin committed:
perf(ivy): speed up ngcc ivy switch processing

Only parse the AST for ngcc ivy switch constants
if the marker is not found in the module text.

mary-poppins commented:
You can preview 29457cc at https://pr25534-29457cc.ngbuilds.io/.

mhevery commented:
http://test/OCL:211042347:BASE:211042348:1535698506772:8183a8d4
